File: components/esp_board_manager/peripherals/periph_pcnt/periph_pcnt.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# PCNT peripheral config parser
VERSION = 'v1.0.0'

# Define valid PCNT configuration values
VALID_PCNT_CHANNEL_EDGE_ACTIONS = [
    'PCNT_CHANNEL_EDGE_ACTION_HOLD',
    'PCNT_CHANNEL_EDGE_ACTION_INCREASE',
    'PCNT_CHANNEL_EDGE_ACTION_DECREASE'
]

VALID_PCNT_CHANNEL_LEVEL_ACTIONS = [
    'PCNT_CHANNEL_LEVEL_ACTION_KEEP',
    'PCNT_CHANNEL_LEVEL_ACTION_INVERSE',
    'PCNT_CHANNEL_LEVEL_ACTION_HOLD'
]

# SOC constants (adjust based on actual SOC)
SOC_PCNT_CHANNELS_PER_UNIT = 2
SOC_PCNT_THRES_POINT_PER_UNIT = 5

def validate_enum_value(value: str, enum_type: str, valid_values: list) -> bool:
    """Validate if the enum value is within the valid range.

    Args:
        value: The enum value to validate
        enum_type: The type of enum for error messages
        valid_values: List of valid values

    Returns:
        True if valid, False otherwise
    """
    if value not in valid_values:
        print(f"❌ YAML VALIDATION ERROR: Invalid {enum_type} value '{value}'!")
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: PCNT configuration')
        print(f"   ❌ Invalid value: '{value}'")
        print(f'   ✅ Valid values: {valid_values}')
        print(f'   ℹ️ Please use one of the valid enum values listed above')
        return False
    return True

def get_enum_value(value: str, default_value: str, enum_type: str, valid_values: list) -> str:
    """Helper function to get enum value with validation.

    Args:
        value: The value from YAML config
        default_value: The default value to use if value is not provided
        enum_type: The type of enum for validation
        valid_values: List of valid values

    Returns:
        The enum value to use
    """
    if not value:
        result = default_value
    else:
        result = value

    # Validate the enum value
    if not validate_enum_value(result, enum_type, valid_values):
        raise ValueError(f"YAML validation error in PCNT peripheral: Invalid {enum_type} value '{result}'. Valid values: {valid_values}")

    return result

def get_includes() -> list:
    """Return list of required include headers for PCNT"""
    return [
        'periph_pcnt.h',
    ]

def parse_pcnt_unit_config(unit_config: dict) -> dict:
    """Parse PCNT unit configuration"""
    unit_cfg = {
        'low_limit': unit_config.get('low_limit', -100),
        'high_limit': unit_config.get('high_limit', 100),
        'intr_priority': unit_config.get('intr_priority', 0)
    }

    # Parse flags
    flags_dict = unit_config.get('flags', {})
    flags_config = {
        'accum_count': int(flags_dict.get('accum_count', False))  # Convert bool to int
    }

    if flags_dict.get('en_step_notify_up') is not None:
        flags_config['en_step_notify_up'] = int(flags_dict.get('en_step_notify_up', False))

    if flags_dict.get('en_step_notify_down') is not None:
        flags_config['en_step_notify_down'] = int(flags_dict.get('en_step_notify_down', False))

    # Add SOC-specific flags if supported
    # Note: These would be conditionally included based on SOC capabilities
    unit_cfg['flags'] = flags_config

    return unit_cfg

def parse_pcnt_filter_config(filter_config: dict) -> dict:
    """Parse PCNT glitch filter configuration"""
    filter_cfg = {
        'max_glitch_ns': filter_config.get('max_glitch_ns', 1000)
    }

    return filter_cfg

def parse_pcnt_channel_list(channel_list: list) -> list:
    """Parse PCNT channel configuration list"""
    parsed_channels = []

    for channel_config in channel_list:
        # Parse channel configuration
        channel_cfg_raw = channel_config.get('channel_config', {})
        flags_raw = channel_cfg_raw.get('flags', {})

        edge_gpio = channel_cfg_raw.get('edge_gpio_num', -1)
        level_gpio = channel_cfg_raw.get('level_gpio_num', -1)

        invert_edge_input = int(flags_raw.get('invert_edge_input', False))
        invert_level_input = int(flags_raw.get('invert_level_input', False))
        virt_edge_io_level = int(flags_raw.get('virt_edge_io_level', 0))
        virt_level_io_level = int(flags_raw.get('virt_level_io_level', 0))
        io_loop_back = int(flags_raw.get('io_loop_back', False))

        flags_dict = {
            'invert_edge_input': invert_edge_input,
            'invert_level_input': invert_level_input,
            'virt_edge_io_level': virt_edge_io_level,
            'virt_level_io_level': virt_level_io_level,
        }

        from generators.utils.idf_version import get_idf_version
        if get_idf_version()[0] < 6:
            flags_dict['io_loop_back'] = io_loop_back
        elif io_loop_back:
            print(f"⚠️ YAML WARNING: 'io_loop_back' flag in PCNT configuration '{channel_config.get('name', 'unknown')}' is deprecated in IDF >= 6.0 and will be ignored.")

        pos_act = get_enum_value(
            channel_config.get('pos_act'),
            'PCNT_CHANNEL_EDGE_ACTION_INCREASE',
            'PCNT positive edge action',
            VALID_PCNT_CHANNEL_EDGE_ACTIONS
        )
        neg_act = get_enum_value(
            channel_config.get('neg_act'),
            'PCNT_CHANNEL_EDGE_ACTION_DECREASE',
            'PCNT negative edge action',
            VALID_PCNT_CHANNEL_EDGE_ACTIONS
        )
        high_act = get_enum_value(
            channel_config.get('high_act'),
            'PCNT_CHANNEL_LEVEL_ACTION_KEEP',
            'PCNT high level action',
            VALID_PCNT_CHANNEL_LEVEL_ACTIONS
        )
        low_act = get_enum_value(
            channel_config.get('low_act'),
            'PCNT_CHANNEL_LEVEL_ACTION_KEEP',
            'PCNT low level action',
            VALID_PCNT_CHANNEL_LEVEL_ACTIONS
        )

        parsed_channels.append({
            'channel_config': {
                'edge_gpio_num': edge_gpio,
                'level_gpio_num': level_gpio,
                'flags': flags_dict
            },
            'pos_act': pos_act,
            'neg_act': neg_act,
            'high_act': high_act,
            'low_act': low_act
        })

    return parsed_channels

def validate_channel_count(channel_count: int, channel_list: list) -> bool:
    """Validate that channel count matches the number of channels in the list"""
    if channel_count != len(channel_list):
        print(f'❌ YAML VALIDATION ERROR: Channel count mismatch!')
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: PCNT configuration')
        print(f'   ❌ channel_count: {channel_count}')
        print(f'   ❌ channel_list length: {len(channel_list)}')
        print(f'   ℹ️ channel_count must match the number of channels in channel_list')
        return False

    if channel_count > SOC_PCNT_CHANNELS_PER_UNIT:
        print(f'❌ YAML VALIDATION ERROR: Too many channels!')
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: PCNT configuration')
        print(f'   ❌ Number of channels: {channel_count}')
        print(f'   ℹ️ Maximum channels per unit: {SOC_PCNT_CHANNELS_PER_UNIT}')
        return False

    return True

def validate_watch_point_count(watch_point_count: int, watch_point_list: list) -> bool:
    """Validate that watch point count matches the number of watch points in the list"""
    if watch_point_count != len(watch_point_list):
        print(f'❌ YAML VALIDATION ERROR: Watch point count mismatch!')
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: PCNT configuration')
        print(f'   ❌ watch_point_count: {watch_point_count}')
        print(f'   ❌ watch_point_list length: {len(watch_point_list)}')
        print(f'   ℹ️ watch_point_count must match the number of watch points in watch_point_list')
        return False

    if watch_point_count > (SOC_PCNT_THRES_POINT_PER_UNIT + 3):
        print(f'❌ YAML VALIDATION ERROR: Too many watch points!')
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: PCNT configuration')
        print(f'   ❌ Number of watch points: {watch_point_count}')
        print(f'   ℹ️ Maximum watch points per unit: {SOC_PCNT_THRES_POINT_PER_UNIT + 3}')
        return False

    return True

def parse(name: str, full_config: dict, peripherals_dict=None) -> dict:
    """Parse PCNT peripheral configuration from YAML"""
    # Get the actual config from the full_config
    config = full_config.get('config', {})

    # Parse unit configuration
    unit_config_raw = config.get('unit_config', {})
    unit_config = parse_pcnt_unit_config(unit_config_raw)

    # Parse filter configuration
    filter_config_raw = config.get('filter_config', {})
    filter_config = parse_pcnt_filter_config(filter_config_raw)

    # Parse channel configuration
    channel_count = config.get('channel_count', 0)
    channel_list_raw = config.get('channel_list', [])

    # Validate channel count
    if not validate_channel_count(channel_count, channel_list_raw):
        raise ValueError(f"Invalid channel configuration for PCNT peripheral '{name}'")

    # Parse channel list
    channel_list = parse_pcnt_channel_list(channel_list_raw)

    # Parse watch points
    watch_point_count = config.get('watch_point_count', 0)
    watch_point_list = config.get('watch_point_list', [])

    # Validate watch point count
    if not validate_watch_point_count(watch_point_count, watch_point_list):
        raise ValueError(f"Invalid watch point configuration for PCNT peripheral '{name}'")

    # Build the main configuration structure
    struct_init = {
        'unit_config': unit_config,
        'filter_config': filter_config,
        'channel_count': channel_count,
        'channel_list': channel_list,
        'watch_point_count': watch_point_count,
        'watch_point_list': watch_point_list
    }

    # Build the result
    result = {
        'struct_type': 'periph_pcnt_config_t',
        'struct_var': f'{name}_cfg',
        'struct_init': struct_init,
        '_channel_count': channel_count,  # Store for code generation
        '_watch_point_count': watch_point_count  # Store for code generation
    }

    return result


def extract_metadata(name: str, raw_config: dict, parse_result: dict, context: dict) -> dict:
    channel_list = parse_result.get('struct_init', {}).get('channel_list', [])
    edge_gpio_nums = []
    level_gpio_nums = []

    for item in channel_list:
        channel_cfg = item.get('channel_config', {})
        edge_gpio_num = channel_cfg.get('edge_gpio_num', -1)
        level_gpio_num = channel_cfg.get('level_gpio_num', -1)

        if isinstance(edge_gpio_num, int) and edge_gpio_num >= 0 and edge_gpio_num not in edge_gpio_nums:
            edge_gpio_nums.append(edge_gpio_num)
        if isinstance(level_gpio_num, int) and level_gpio_num >= 0 and level_gpio_num not in level_gpio_nums:
            level_gpio_nums.append(level_gpio_num)

    io = {}
    if edge_gpio_nums:
        io['edge_gpio_num'] = edge_gpio_nums
    if level_gpio_nums:
        io['level_gpio_num'] = level_gpio_nums

    return {'io': io}

File: components/esp_board_manager/peripherals/periph_pcnt/periph_pcnt.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <stdlib.h>
#include "esp_log.h"
#include "periph_pcnt.h"

static const char *TAG = "PERIPH_PCNT";

int periph_pcnt_init(void *cfg, int cfg_size, void **periph_handle)
{
    if (!cfg || !periph_handle || cfg_size < sizeof(periph_pcnt_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    esp_err_t err = ESP_FAIL;
    periph_pcnt_config_t *config = (periph_pcnt_config_t *)cfg;
    periph_pcnt_handle_t *handle = (periph_pcnt_handle_t *)calloc(1, sizeof(periph_pcnt_handle_t));
    if (handle == NULL) {
        ESP_LOGE(TAG, "Failed to allocate memory for periph_pcnt_handle_t");
        return -1;
    }

    /* create unit */
    err = pcnt_new_unit(&config->unit_config, &handle->pcnt_unit);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Creat pcnt unit failed: %s", esp_err_to_name(err));
        free(handle);
        return -1;
    }

    /* set glitch filter */
    err = pcnt_unit_set_glitch_filter(handle->pcnt_unit, &config->filter_config);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Set pcnt glitch filter failed: %s", esp_err_to_name(err));
        pcnt_del_unit(handle->pcnt_unit);
        free(handle);
        return -1;
    }

    /* create channels */
    uint8_t ch_count = config->channel_count;
    if (ch_count > SOC_PCNT_CHANNELS_PER_UNIT) {
        ESP_LOGW(TAG, "Channel count %d exceeds max %d, limiting to max", ch_count, SOC_PCNT_CHANNELS_PER_UNIT);
        ch_count = SOC_PCNT_CHANNELS_PER_UNIT;
    }
    for (uint8_t i = 0; i < ch_count; i++) {
        periph_pcnt_channel_config_t *ch_cfg = &config->channel_list[i];
        pcnt_channel_handle_t ch = NULL;
        err = pcnt_new_channel(handle->pcnt_unit, &ch_cfg->channel_config, &ch);
        if (err != ESP_OK) {
            ESP_LOGE(TAG, "Apply pcnt channel %d failed: %s", i, esp_err_to_name(err));
            /* cleanup created channels + unit */
            for (uint8_t j = 0; j < i; j++) {
                if (handle->channels[j]) {
                    pcnt_del_channel(handle->channels[j]);
                    handle->channels[j] = NULL;
                }
            }
            pcnt_del_unit(handle->pcnt_unit);
            free(handle);
            return -1;
        }
        handle->channels[i] = ch;

        /* set edge and level actions if provided */
        pcnt_channel_set_edge_action(ch, ch_cfg->pos_act, ch_cfg->neg_act);
        pcnt_channel_set_level_action(ch, ch_cfg->high_act, ch_cfg->low_act);
    }

    /* add watch points if any */
    if (config->watch_point_count > 0) {
        uint8_t wp_cnt = config->watch_point_count;
        if (wp_cnt > SOC_PCNT_THRES_POINT_PER_UNIT + 3) {
            wp_cnt = SOC_PCNT_THRES_POINT_PER_UNIT + 3;
            ESP_LOGW(TAG, "Watch point count exceeds max %d, limiting to max", SOC_PCNT_THRES_POINT_PER_UNIT + 3);
        }
        for (uint8_t i = 0; i < wp_cnt; i++) {
            pcnt_unit_add_watch_point(handle->pcnt_unit, config->watch_point_list[i]);
        }
    }

    *periph_handle = (void *)handle;
    ESP_LOGI(TAG, "Periph pcnt init success: unit=%p", handle->pcnt_unit);
    return 0;
}

int periph_pcnt_deinit(void *periph_handle)
{
    if (periph_handle == NULL) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }
    periph_pcnt_handle_t *handle = (periph_pcnt_handle_t *)periph_handle;

    esp_err_t err = ESP_FAIL;
    /* delete channels */
    for (size_t i = 0; i < SOC_PCNT_CHANNELS_PER_UNIT; i++) {
        if (handle->channels[i]) {
            err = pcnt_del_channel(handle->channels[i]);
            if (err != ESP_OK) {
                ESP_LOGE(TAG, "Failed to delete channel %d: %s", i, esp_err_to_name(err));
                return -1;
            }
            handle->channels[i] = NULL;
        }
    }

    /* delete unit */
    if (handle->pcnt_unit) {
        err = pcnt_del_unit(handle->pcnt_unit);
        if (err != ESP_OK) {
            ESP_LOGE(TAG, "Failed to delete pcnt unit: %s", esp_err_to_name(err));
            return -1;
        }
        handle->pcnt_unit = NULL;
    }

    free(handle);
    ESP_LOGI(TAG, "Periph pcnt deinit success");
    return 0;
}
